Fixed top toolbar: adjust height to ensure parent's bottom border displays

ramonjd opened the pull request:
Resolves https://github.com/WordPress/gutenberg/issues/54093

<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Reduce the height of the fixed block toolbar by 1px. 

## Why and how?
When the block toolbar is fixed, account for parent's bottom border. The [parent is also set at 60px high](https://github.com/WordPress/gutenberg/blob/57c4a6ea53038423d8fff35e8d65c1ff782895b8/packages/edit-site/src/components/header-edit-mode/style.scss#L4), including the border so to make it fit we have to squish the toolbar to 59px. 

Another approach would be to give the block toolbar a bottom border, but seems fragile since if the toolbar is pushed out in a vertical direction the double border would show.

## Testing Instructions

In the site editor, enable the fixed block tookbar in the settings panel

<img width="365" alt="Screenshot 2023-09-03 at 1 00 59 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/3c5c566b-2235-4b3c-83e0-55deef18c806">

Select a block, e.g., the site title block. The bottom border on the fixed toolbar should display properly.

Check the post editor as well.

| Before | After |
| ------------- | ------------- |
| <img width="300" alt="Screenshot 2023-09-03 at 12 40 56 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/088f4221-919d-4c7e-9907-b3f51c730d95">  | <img width="300" alt="Screenshot 2023-09-03 at 12 40 40 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/5fac200a-4863-4406-9491-005fe4627010">  |


Note: you might see a horizontal scroll bar, this is also a feature on trunk too:


https://github.com/WordPress/gutenberg/assets/6458278/f780e62f-6c2d-4872-bd00-ce171b83231f

ramonjd committed:
Account for parent's bottom border. The parent is also 60px high, including the border so that makes 59px

t-hamano reviewed (commented):
Thanks for the PR!

The approach of minus 1 makes sense to me. Just to be sure, I would like to hear from @draganescu, who is familiar with toolbars.

Tested on Windows 11:


https://github.com/WordPress/gutenberg/assets/54422211/be696148-f9a2-4ebc-af47-7eeee0a2a8b6

draganescu reviewed (commented):
The code to add height was added to allow the scrollbar to appear "inside" when the items overflow. On OSX Firefox and Safari the problem does not appear. Is it a Windows browser problem?

ramonjd commented:
> The code to add height was added to allow the scrollbar to appear "inside" when the items overflow. On OSX Firefox and Safari the problem does not appear. Is it a Windows browser problem?

Do mean this?

https://github.com/WordPress/gutenberg/assets/6458278/f780e62f-6c2d-4872-bd00-ce171b83231f

I thought it was a UI bug, that's why I mentioned it. I could see it on trunk on FF/Chrome and Safari. 

Good to know it isn't a bug! 😄 

I only added a note in the description just in case folks testing thought it was caused by the changed in this PR (it's not).

This PR just tries to account for the bottom border and squishes the child element height a bit. Do you reckon it's okay?

draganescu commented:
The problem does not manifest in the post editor.

<img width="473" alt="Screenshot 2023-09-06 at 10 20 13" src="https://github.com/WordPress/gutenberg/assets/107534/0534d87a-611f-4e5c-b63f-6d031ce61902">

It's only visible in the site editor. Does that mean that maybe the site editor does something w/ header CSS?

andrewserong reviewed (approved):
This is looking good and fixes the issue of the bottom border.

For the horizontal scrollbar issue, I can only reproduce it if a certain number of items are in the toolbar. So for example with the Heading block selected it'll appear at a certain viewport width, but not with the Paragraph block:

![2023-09-07 16 46 11](https://github.com/WordPress/gutenberg/assets/14988353/850a7adf-3d0e-400d-ad38-cabdb7042f5b)

That does look like a bug to me — perhaps some of the values for when to show or hide elements in the header at different breakpoints might need to be adjusted slightly? Or, maybe when there's a horizontal scrollbar, a different design could be used that gracefully fades the right edge of the toolbar — or moves toolbar items to a more menu? All that could require some design work to figure out, though 🤔

In any case, the horizontal scrollbar appears consistently for me between this PR and trunk, so I think it sounds like something that could be investigated separately.

LGTM! ✨

draganescu commented:
@andrewserong what horizontal scrollbar issue 😁 ? I am a bit lost here.

I don't want to block the PR - it's reducing height by 1px it's fine, but I have to highlight that the missing border happens only in the site editor not in the post editor so maybe the -1px is not the solution, but instead the site editor does something that pushes the toolbar down, maybe IDK.

t-hamano commented:
I think the reason the bottom border missing issue doesn't occur in the Post Editor is because the header is actually **61px** height. In other words, the bottom border is applied to the parent element of an element with a height of **60px**.

![Post Editor](https://github.com/WordPress/gutenberg/assets/54422211/a8d05f93-8723-4aba-a960-1ac40ad3a257)

On the other hand, in the Site Editor, the bottom border is applied directly to the element with a height of 60px, which is why I think this problem occurs.

![Site Editor](https://github.com/WordPress/gutenberg/assets/54422211/44faba57-8f40-4f09-9295-e25238564491)

Should the header height be 60px or 61px? 🤔

ramonjd commented:
Thanks @t-hamano and @draganescu for thinking about this. 

> On the other hand, in the Site Editor, the bottom border is applied directly to the element with a height of 60px, which is why I think this problem occurs.

And `interface-interface-skeleton__header` in the site editor is empty 🤷🏻 

I've tested this, and it looks like it will work (will test more thoroughly tomorrow)

```diff
diff --git a/packages/edit-site/src/components/editor/style.scss b/packages/edit-site/src/components/editor/style.scss
index 26403858ff..b0d5d3bc74 100644
--- a/packages/edit-site/src/components/editor/style.scss
+++ b/packages/edit-site/src/components/editor/style.scss
@@ -8,7 +8,7 @@
 	}
 
 	.interface-interface-skeleton__header {
-		border: 0;
+	  border-bottom: $border-width solid $gray-200;
 	}
 }
 
diff --git a/packages/edit-site/src/components/header-edit-mode/style.scss b/packages/edit-site/src/components/header-edit-mode/style.scss
index 26b1716a28..d6e93071b8 100644
--- a/packages/edit-site/src/components/header-edit-mode/style.scss
+++ b/packages/edit-site/src/components/header-edit-mode/style.scss
@@ -9,7 +9,6 @@ $header-toolbar-min-width: 335px;
 	box-sizing: border-box;
 	width: 100%;
 	justify-content: space-between;
-	border-bottom: $border-width solid $gray-200;
 	padding-left: $header-height;
 
 	.edit-site-header-edit-mode__start {

```

> I don't want to block the PR - it's reducing height by 1px it's fine, but I have to highlight that the missing border happens only in the site editor not in the post editor so maybe the -1px is not the solution, but instead the site editor does something that pushes the toolbar down, maybe IDK.

You're right, sorry I didn't understand at first, but it would be good to harmonize them to prevent hacks like this one.

andrewserong commented:
> @andrewserong what horizontal scrollbar issue 😁 ? I am a bit lost here.

My apologies @draganescu, I think I'm repeating the discussion with Ramon from https://github.com/WordPress/gutenberg/pull/54131#issuecomment-1707435130 — as in that comment, I thought the horizontal scrollbar was a UI bug rather than intentional at intermediate breakpoints in the post and site editors. Because the horizontal scrollbar causes the buttons to jump up vertically and no longer be aligned in a single row (the baseline of all icons in the header), it looked unfinished to me, so I wondered if it could use some more design work to come up with something that feels a little smoother, without it jumping up vertically?

That's not really related to this PR, though, and not a worry at all if folks are happy with the current behaviour!

ramonjd commented:
So, this goes pretty deep, and I've spent way too much time on this 🤣 

The site editor is set up differently. I'm not sure why. I suspect it's the complexity of view/edit modes and the ever-present WP logo.

Anyway, `.edit-site-header-edit-mode` does most of the styling work. 

But in the post editor `.interface-interface-skeleton__header` does it. 

And as we know, `.interface-interface-skeleton__header` is 100% empty on the site editor. 😄 

So I tried styling `.interface-interface-skeleton__header` in the site editor, but that means overriding `.edit-site-header-edit-mode`, but only sometimes (in edit mode) and at certain break points due to mobile view.

Here's the diff, whose effect is pretty much the same as the current state of this PR (reducing to 59px). 

```diff
diff --git a/packages/edit-site/src/components/editor/style.scss b/packages/edit-site/src/components/editor/style.scss
index 26403858ff..b5e31efb67 100644
--- a/packages/edit-site/src/components/editor/style.scss
+++ b/packages/edit-site/src/components/editor/style.scss
@@ -8,7 +8,13 @@
 	}
 
 	.interface-interface-skeleton__header {
-		border: 0;
+		border-bottom: 0;
+		.edit-site-layout.is-full-canvas.is-edit-mode & {
+			// Only show border in site editor edit mode.
+			@include break-medium() {
+				border-bottom: $border-width solid $gray-200;
+			}
+		}
 	}
 }
 
diff --git a/packages/edit-site/src/components/header-edit-mode/style.scss b/packages/edit-site/src/components/header-edit-mode/style.scss
index 26b1716a28..11e7fd5d0d 100644
--- a/packages/edit-site/src/components/header-edit-mode/style.scss
+++ b/packages/edit-site/src/components/header-edit-mode/style.scss
@@ -9,8 +9,11 @@ $header-toolbar-min-width: 335px;
 	box-sizing: border-box;
 	width: 100%;
 	justify-content: space-between;
-	border-bottom: $border-width solid $gray-200;
 	padding-left: $header-height;
+  	border-bottom: $border-width solid $gray-200;
+	@include break-medium() {
+		border-bottom: 0;
+	}
 
 	.edit-site-header-edit-mode__start {
 		display: flex;
diff --git a/packages/edit-site/src/components/layout/style.scss b/packages/edit-site/src/components/layout/style.scss
index 11c7bdeeaf..8ac7659027 100644
--- a/packages/edit-site/src/components/layout/style.scss
+++ b/packages/edit-site/src/components/layout/style.scss
@@ -178,6 +178,10 @@
 	.edit-site-layout.is-full-canvas.is-edit-mode & {
 		border-bottom-color: $gray-200;
 		transition: border-bottom-color 0.15s 0.4s ease-out;
+		@include break-medium() {
+			// Height to match the post editor logo.
+			height: $header-height + $border-width;
+		}
 	}
 
 	&:hover,

```

 
I'm inclined to prefer a one-liner than all this breakpoint voodoo.

andrewserong commented:
Oh, good sleuthing! Yeah, I'd never noticed before that the horizontal line in the post editor sits beneath the logo, but in the site editor it is at the same level as the bottom pixels of the logo area:

### Post editor

<img width="360" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/7b5632f9-1223-4357-8892-e8fa8cbe1d0a">

### Site editor

<img width="326" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/c13ec09b-c7c4-4cf4-82c8-0de3ee6c24c4">

I imagine the site editor's styling is probably the desired longer-term direction?

> I'm inclined to prefer a one-liner than all this breakpoint voodoo.

Yeah, I'd be inclined to go with the one-liner, too, even if it feels a little hacky to have to include the `- 1` calculation, it seems a lot neater than adding in the overrides for all those breakpoints.

draganescu commented:
The border _under_ the logo should not be the preffered pattern, I think? cc @jasmussen 😁

ramonjd commented:
> Yeah, I'd never noticed before that the horizontal line in the post editor sits beneath the logo, but in the site editor it is at the same level as the bottom pixels of the logo area:

Thanks for checking @andrewserong 

Just to confirm, you're looking at trunk, right?

On this PR, it's "standardized" to sit underneath the logo (the height jump also appears on trunk and is a separate issue I reckon)

https://github.com/WordPress/gutenberg/assets/6458278/ab0d0ad2-e277-4ed2-8476-f81fa9ae7d7c

andrewserong commented:
> Thanks for checking @andrewserong
> Just to confirm, you're looking at trunk, right?

Yes, I was testing on `trunk`. Re-testing this morning, and I think I got it the wrong way around last Friday — today, I see the border sits below the logo box in the site editor, but not below in the post editor (on trunk).

With this PR, I see the border beneath the logo box in both the post and site editors, so it's more consistent than on trunk 👍

> The border under the logo should not be the preffered pattern, I think? cc @jasmussen 😁

Good question, whichever way we go with, it'd be good for it to be consistent between the two editors. Might be worth getting some design feedback on the preferred approach here?

Thanks again for all the discussion here, I know it's tricky when we're just dealing with a one-liner!

ramonjd commented:
What do folks think about merging this, since it address the main issue, and moving discussion to a new issue?

I see a couple of site editor header issues that could be addressed:

- Is it important to maintain parity with the post editor? (height, collapsed states, styles) 
- If so, how could we refactor?


cc @WordPress/gutenberg-design

jasmussen commented:
Catching up, sorry I've been AFK. Thank you for the attention to detail here, folks.

I seem to recall either @richtabor or @jameskoster intentionally moving the border to underlap the logo button. OR, making the border color semi-transparent so that it better blends in. I could go either way here, but I also think that more important than which direction we go here, is that the metrics are shared between the post and site editors, ideally at a component level. Right now there's a ton of drift, and I'd personally be willing to accept some detours if it means we can clean up the code.

jameskoster commented:
Agree that parity between editors is important. 

The 59px height on the top toolbar is pushing the icons inside (and other visual details like the separators) out of alignment with the main tools:

<img width="948" alt="Screenshot 2023-09-11 at 16 08 46" src="https://github.com/WordPress/gutenberg/assets/846565/e293ebe3-e7d1-420a-a781-126b2b5b5d40">

In the screenshot above notice how the main document tools align nicely with the red guidelines while the elements in the top toolbar are a couple of pixels too high. Adding `padding-top: 1px` to `.block-editor-block-contextual-toolbar` fixes this.

---

If you twisted my arm I would actually push for a **64px** internal height as that allows all the other values to align with the 4px baseline:

<img width="752" alt="Screenshot 2023-09-11 at 15 59 25" src="https://github.com/WordPress/gutenberg/assets/846565/6cc0c51c-5e68-4a05-b404-cef73ff8b595">

It also makes the site icon exactly half-size in site view which is nice and neat. 

Appreciate this would be more than a single-line change, but I'd be happy to work with someone on this. Otherwise the small bump in padding as suggested above would probably make this change acceptable.

ramonjd committed:
Moving site editor specific styles to edit-site package
Adding 1px padding top to fixed toolbar to compensate for the change in block tool icons alignment

ramonjd commented:
Thanks for the feedback @jasmussen and @jameskoster 

> Adding padding-top: 1px to .block-editor-block-contextual-toolbar fixes this.

Oh good call. I've updated this PR, and also moved the sites to the edit-site package in order to isolate them further.

I'm reading the general vibe 😄 and sounds like we should move the refactor of the site editor header discussion to another issue. I can create it, though I don't really have the bandwidth this week to work on it. So maybe some other folks might pick it up in the meantime. 

I'm still a bit uncertain about the fate of the [original bug report](https://github.com/WordPress/gutenberg/issues/54093) however, but if folks are happy for the block tools bottom border to remain cut off until the wider design issues are addressed I'm happy to close this. 

My opinion remains that we should merge it, because it removes an eye sore with little tech debt overhead.

ramonjd commented:
I've created an issue for a follow up task to try to unify the headers' markup/styles:

- https://github.com/WordPress/gutenberg/issues/54367

andrewserong commented:
> My opinion remains that we should merge it, because it removes an eye sore with little tech debt overhead.

Same here. I just gave this a re-test and it's still working nicely for me, I like the approach of restricting the rule to the edit-site package for now, too, so that it's more of a local fix 👍

To my eyes it's looking nice and lined up now, but might be good to double-check with @jameskoster for a more detailed look at the pixels 🙂

jasmussen commented on packages/edit-site/src/components/block-editor/style.scss:
```diff
@@ -173,3 +173,14 @@
+		// Minus one to account for the bottom border of the .is-fixed parent.
+		height: $header-height - 1;
+		// Ensure the block tools icons remain aligned with the main toolbar items.
+		padding-top: 1px;
```
Can replace both `1` and  `1px` with `$border-width`.

ramonjd replied:
Nice. Done!

jasmussen commented:
Indeed, there is work to be done. See also #46734.

However I agree, the top toolbar eye-sore is bad and this is worth fixing. I left a small comment on using a sass variable, and _the other important thing_ is that post and site editors should be the same. Let's come back to this one, so long as we do that, fine to move forward.

ramonjd committed:
Update style.scss

use $border-width for 1px

jameskoster commented:
✅ All nicely lined up now 

<img width="962" alt="Screenshot 2023-09-12 at 13 03 31" src="https://github.com/WordPress/gutenberg/assets/846565/f15d7f44-ab27-40ce-97fc-ae99168eafff">

richtabor commented:
> I seem to recall either @richtabor or @jameskoster intentionally moving the border to underlap the logo button. OR, making the border color semi-transparent so that it better blends in. I could go either way here, but I also think that more important than which direction we go here, is that the metrics are shared between the post and site editors, ideally at a component level. 

An oldie, but I like the approach I explored in https://github.com/WordPress/gutenberg/pull/42282, where the borders don't clash with the background color (or anything really). The borders are there when you need them, but not when you don't. 


![](https://user-images.githubusercontent.com/1813435/178070608-73f71960-3a05-4b57-9448-a0d0df5cd7ba.png)

ramonjd commented:
Thanks again for everyone's help on this PR. 

It appears that https://github.com/WordPress/gutenberg/pull/54356 has resolved the main issue (https://github.com/WordPress/gutenberg/issues/54093) overnight. 🎉 😆 

Here's the current stage of trunk:


https://github.com/WordPress/gutenberg/assets/6458278/c036697c-7e1b-4c3b-810a-902f53f11814

The Site Editor header is still 1px smaller in height, but the border is now visible.

I'll trash this PR and make a note on the issue.

ramonjd commented:
Fixed in https://github.com/WordPress/gutenberg/pull/54356
